File: skyspark-scala/src/test/scala/com/akikanellis/skyspark/scala/api/algorithms/bnl/LocalSkylineCalculatorTest.scala
package com.akikanellis.skyspark.scala.api.algorithms.bnl

import com.akikanellis.skyspark.scala.api.algorithms.Point
import com.akikanellis.skyspark.scala.test_utils.{SparkAddOn, UnitSpec}
import org.mockito.Mockito._

class LocalSkylineCalculatorTest extends UnitSpec with SparkAddOn {
  var bnlAlgorithm: BnlAlgorithm = _
  var localSkylineCalculator: LocalSkylineCalculator = _

  before {
    bnlAlgorithm = mockForSpark[BnlAlgorithm]
    localSkylineCalculator = new LocalSkylineCalculator(bnlAlgorithm)
  }

  "A set of points with their flags" should "produce the set of local skylines per key" in withSpark { sc =>
    val flagPointsSeq = Seq(
      (Flag(true, true), Point(5.9, 4.6)),
      (Flag(true, true), Point(6.9, 5.6)),
      (Flag(true, false), Point(5.0, 4.1)),
      (Flag(true, false), Point(5.9, 4.0)))
    val pointsWithExpectedLocalSkylines = Seq(
      (Seq(Point(5.9, 4.6), Point(6.9, 5.6)), Seq(Point(5.9, 4.6))),
      (Seq(Point(5.0, 4.1), Point(5.9, 4.0)), Seq(Point(5.0, 4.1), Point(5.9, 4.0))))
    whenAPointGroupIsToBeComputedReturnSkylines(pointsWithExpectedLocalSkylines)
    val flagPoints = sc.parallelize(flagPointsSeq)
    val expectedLocalSkylinesWithFlags = Seq(
      (Flag(true, true), Point(5.9, 4.6)),
      (Flag(true, false), Point(5.0, 4.1)),
      (Flag(true, false), Point(5.9, 4.0)))

    val localSkylinesWithFlags = localSkylineCalculator.computeLocalSkylines(flagPoints).collect

    localSkylinesWithFlags should contain theSameElementsAs expectedLocalSkylinesWithFlags
  }

  private def whenAPointGroupIsToBeComputedReturnSkylines(pointsWithExpectedLocalSkylines: Seq[(Seq[Point], Seq[Point])]) = {
    pointsWithExpectedLocalSkylines.foreach {
      case (points, skylines) => when(bnlAlgorithm.computeSkylinesWithoutPreComparison(points)).thenReturn(skylines)
    }
  }
}

File: skyspark-java/src/test/java/com/akikanellis/skyspark/api/algorithms/bnl/LocalSkylineCalculatorTest.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import com.akikanellis.skyspark.api.test_utils.SparkContextRule;
import org.apache.spark.api.java.JavaPairRDD;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import scala.Tuple2;

import java.util.List;

import static com.akikanellis.skyspark.api.test_utils.Mockitos.mockForSpark;
import static com.akikanellis.skyspark.api.test_utils.assertions.MoreAssertions.assertThat;
import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static org.mockito.Mockito.when;

public class LocalSkylineCalculatorTest {
    @Rule public SparkContextRule sc = new SparkContextRule();

    private BnlAlgorithm bnlAlgorithm;
    private LocalSkylineCalculator localSkylineCalculator;

    @Before public void beforeEach() {
        bnlAlgorithm = mockForSpark(BnlAlgorithm.class);
        localSkylineCalculator = new LocalSkylineCalculator(bnlAlgorithm);
    }

    @Test public void producingLocalSkylines_producesSkylinesPerKey() {
        List<Tuple2<Flag, Point>> flagPointsList = asList(
                new Tuple2<>(new Flag(true, true), new Point(5.9, 4.6)),
                new Tuple2<>(new Flag(true, true), new Point(6.9, 5.6)),
                new Tuple2<>(new Flag(true, false), new Point(5.0, 4.1)),
                new Tuple2<>(new Flag(true, false), new Point(5.9, 4.0))
        );
        whenAPointGroupIsToBeComputedReturnSkylines();
        JavaPairRDD<Flag, Point> flagPoints = sc.parallelizePairs(flagPointsList);
        List<Tuple2<Flag, Point>> expectedLocalSkylinesWithFlags = asList(
                new Tuple2<>(new Flag(true, true), new Point(5.9, 4.6)),
                new Tuple2<>(new Flag(true, false), new Point(5.0, 4.1)),
                new Tuple2<>(new Flag(true, false), new Point(5.9, 4.0))
        );

        List<Tuple2<Flag, Point>> localSkylinesWithFlags
                = localSkylineCalculator.computeLocalSkylines(flagPoints).collect();

        assertThat(localSkylinesWithFlags).containsOnlyElementsOf(expectedLocalSkylinesWithFlags);
    }

    private void whenAPointGroupIsToBeComputedReturnSkylines() {
        when(bnlAlgorithm.computeSkylinesWithoutPreComparison(asList(new Point(5.9, 4.6), new Point(6.9, 5.6))))
                .thenReturn(singletonList(new Point(5.9, 4.6)));
        when(bnlAlgorithm.computeSkylinesWithoutPreComparison(asList(new Point(5.0, 4.1), new Point(5.9, 4.0))))
                .thenReturn(asList(new Point(5.0, 4.1), new Point(5.9, 4.0)));
    }

}